### PR TITLE
Fix bug where we kept re-requesting a remote server's key repeatedly. 

### PR DESCRIPTION
**Pull request by erikjohnston**
We have two tables:
1. `server_keys_json` which stores the raw JSON that we got, and is unique on `(server_name, key_json, from_server)`.
2. `server_signature_keys` which stores the JSON fields as columns, and is unique on `(server_name, key_json)`

When fetching keys from the store in `Keyring`, we used the first table. However, the fetch function didn't correctly account for multiple rows per server / key ID, and so would return whichever row happened to be returned last by Postgres. (This is often the most recent row, but also often not). 

The fix here is: a) fix the read path to return the most recently added key JSON, and b) use the `server_signature_keys` instead for `Keyring` as we don't need the full JSON anyway.

### CODE REVIEW

File: synapse/crypto/keyring.py [hunk 1]
```diff
@@ -511,7 +511,7 @@ async def _fetch_keys(
             for key_id in queue_value.key_ids
         )
 
-        res = await self.store.get_server_keys_json(key_ids_to_fetch)
+        res = await self.store.get_server_signature_keys(key_ids_to_fetch)
```
Comment (clokep) [inline synapse/crypto/keyring.py:514]: This essentially reverts #15417 and will cause large amounts of federation traffic. See #15171.

**Review by clokep — CHANGES_REQUESTED** [review]

**Comment by clokep** [comment]
See also #15463.

File: synapse/crypto/keyring.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/crypto/keyring.py:514]: Oh boo, for some reason I thought that the both perspective and direct key fetcher persisted in there :facepalm:

**Comment by erikjohnston** [comment]
@clokep I've reverted the second set change. I've rebased (sorry about that) to only be the basic fix

File: synapse/crypto/keyring.py (hunk 1 above)
Comment (clokep) [reply synapse/crypto/keyring.py]: Opposite, see https://github.com/matrix-org/synapse/issues/15171#issuecomment-1464224490 they both store into `server_keys_json`.

**Pull request merged**